### PR TITLE
fix new inventory azure_rm.py

### PR DESCRIPTION
**Pull request by yungezz**
##### SUMMARY

fix issues in new azure_rm.py, including https://github.com/ansible/ansible/issues/47268
1. instance_view and network interface api response body is not in `responses`, but `value`.
2. network interface may hasn't ip when dynamic allocated.
3. add name and x-ms-client-request-id to request and header, other batch size > 20 calls fail.
4. response not matching handler queued

tested against azure subs with more than 20 vms.

##### ISSUE TYPE

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by yungezz** [comment]
@zikalino  @yuwzho pls help to review.

**Comment by yungezz** [comment]
https://github.com/ansible/ansible/issues/47268

File: lib/ansible/plugins/inventory/azure_rm.py [hunk 1]
```diff
@@ -514,8 +524,9 @@ def _on_instanceview_response(self, vm_instanceview_model):
                                  for s in vm_instanceview_model.get('statuses', []) if self._powerstate_regex.match(s.get('code', ''))), 'unknown')
 
     def _on_nic_response(self, nic_model, is_primary=False):
-        nic = AzureNic(nic_model=nic_model, inventory_client=self._inventory_client, is_primary=is_primary)
-        self.nics.append(nic)
+        if 'type' in nic_model and nic_model['type'] == 'Microsoft.Network/networkInterfaces':
```
Comment (yuwzho) [inline lib/ansible/plugins/inventory/azure_rm.py]: Can you add a comment for the `else` scenario?
The `if` conditional check can be written as `if nic_model.get('type') == 'Microsoft.Network/networkInterfaces'`

File: lib/ansible/plugins/inventory/azure_rm.py [hunk 2]
```diff
@@ -499,7 +509,7 @@ def hostvars(self):
                 pip_id = ipc['properties'].get('publicIPAddress', {}).get('id')
                 if pip_id:
                     pip = nic.public_ips[pip_id]
-                    new_hostvars['public_ipv4_addresses'].append(pip._pip_model['properties']['ipAddress'])
+                    new_hostvars['public_ipv4_addresses'].append(pip._pip_model['properties'].get('ipAddress', None))
```
Comment (yuwzho) [inline lib/ansible/plugins/inventory/azure_rm.py:515]: still append None?

Comment (yungezz) [reply lib/ansible/plugins/inventory/azure_rm.py:515]: follow below fqdn

File: lib/ansible/plugins/inventory/azure_rm.py (hunk 1 above)
Comment (yungezz) [reply lib/ansible/plugins/inventory/azure_rm.py]: resolved

File: lib/ansible/plugins/inventory/azure_rm.py [hunk 3]
```diff
@@ -380,7 +382,7 @@ def _process_queue_batch(self):
                     query_parameters = {'api-version': item.api_version}
                     req = self._client.get(item.url, query_parameters)
 
-                    batch_requests.append(dict(httpMethod="GET", url=req.url))
+                    batch_requests.append(dict(httpMethod="GET", url=req.url, name="ansible" + str(random.randint(1, 1000000))))
```
Comment (AlanCoding) [inline lib/ansible/plugins/inventory/azure_rm.py]: It seems strange to see `random.randint(1, 1000000)`. Obviously, I'd expect that to lead to a 1-in-million collision of names. Why wouldn't you just use alternatives:

 - a simple iterator within the loop where requests are made
 - just call uuid4() to get a globally unique string

**Comment by AlanCoding** [comment]
Hello, I am testing inventory plugins for transitioning AWX and Ansible Tower to use them. When testing the Azure RM plugin with our test infrastructure I hit https://github.com/ansible/ansible/issues/47268

I have checked out this branch, rebased it to recent `devel`, and re-tested with my example and infrastructure. I can confirm this resolves the error, and allows `ansible-inventory` to return JSON content of the inventory as expected.

File: lib/ansible/plugins/inventory/azure_rm.py (hunk 3 above)
Comment (yungezz) [reply lib/ansible/plugins/inventory/azure_rm.py]: resolved

**Comment by yungezz** [comment]
rerady_for_review

**Review by yuwzho — APPROVED** [review]

**Comment by Fred-sun** [comment]
@zikalino @nitzmahone

**Comment by rdtechie** [comment]
I've installed the latest devel version of Ansible on a clean system:
```
ansible --version

ansible 2.8.0.dev0
  config file = /home/aholdvmadmin/rdi-temp/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Sep 12 2018, 05:31:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
```
 and patched azure_rm.py with the code in this commit. I'm faced with the next error when executing:

`ansible-inventory -i cloud.azure_rm.yml --graph -vvv`

```
[WARNING]:  * Failed to parse /home/ansible/cloud.azure_rm.yml with azure_rm plugin: 'value'

  File "/usr/lib/python2.7/site-packages/ansible/inventory/manager.py", line 270, in parse_source plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/inventory/azure_rm.py", line 236, in parse self._get_hosts()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/inventory/azure_rm.py", line 293, in _get_hosts self._process_queue_batch()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/inventory/azure_rm.py", line 409, in _process_queue_batch item.handler(r['content'], **item.handler_args)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/inventory/azure_rm.py", line 355, in _on_vm_page_response for h in response['value']:

 [WARNING]: Unable to parse /home/ansible/cloud.azure_rm.yml as an inventory source
```

Contents of the inventory file:
```
plugin: azure_rm
```

Additional note: I'm getting only around 14 VM's as a result back when I specify all my resource groups in the inventory file, and I have 70+ VM's in the subscription.

**Comment by yungezz** [comment]
@rdtechie i'm trying to reproduce the issue.

**Comment by AlanCoding** [comment]
> [WARNING]:  * Failed to parse /home/ansible/cloud.azure_rm.yml with azure_rm plugin: 'value'

This error reporting is a bug in Ansible (see details in failed fix attempt https://github.com/ansible/ansible/pull/49964), which comes down to a KeyError, looking up "value" in a dictionary in this case.

**Comment by yungezz** [comment]
@rdtechie  i add some defensive code to check `value` key exist in vm response. Could you have a try? for only part  of your vm returned, could you pls check status of those not showing vms? default host filters is  powerstate!=running and provisioning_state!=succeeded.  pls specify customized default_host_filters to include all.

**Comment by yungezz** [comment]
ready_for_review

**Review by AlanCoding — APPROVED** [review]

**Comment by yungezz** [comment]
ready_for_review

**Comment by yungezz** [comment]
ready_for_review

**Comment by yungezz** [comment]
kindly ping @nitzmahone

**Comment by AlanCoding** [comment]
I pulled the most recent version of this branch, and re-tested with our Azure infrastructure. It is working, and still fixes the bug as described earlier.

**Comment by AlanCoding** [comment]
shipit

**Comment by AlanCoding** [comment]
You might want to add a changelog fragment, because this is a fix for a blocking bug, I'd anticipate there may be requests to backport it.

**Comment by yungezz** [comment]
@AlanCoding thank you very much, will add changelog for the change. after create backport pr after this one got merged into devel.

**Comment by AlanCoding** [comment]
Thanks for adding that. The difficulty of backporting may vary with the release branch, so it's an open question of how far back it should be taken. Ping me in the PR and I'll do the same basic functional testing for the versions you attempt.

@nitzmahone is there anything else needed to get this past the finish line.

**Review by nitzmahone — APPROVED** [review]

**Comment by nitzmahone** [comment]
This one was new for 2.7, so only needs one backport.

**Comment by pmarques** [comment]
I'm getting another error now:
```
[WARNING]:  * Failed to parse /workspace/inventory/azr-tst.azure_rm.yml with auto plugin: list indices must be integers or slices, not str                                                                                                                                                                              

  File "/usr/local/lib/python3.7/site-packages/ansible/inventory/manager.py", line 270, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/usr/local/lib/python3.7/site-packages/ansible/plugins/inventory/auto.py", line 55, in parse
    plugin.parse(inventory, loader, path, cache=cache)
  File "/workspace/inventory_plugins/azure_rm.py", line 254, in parse
    self._get_hosts()
  File "/workspace/inventory_plugins/azure_rm.py", line 341, in _get_hosts
    self._process_queue_batch()
  File "/workspace/inventory_plugins/azure_rm.py", line 435, in _process_queue_batch
    result = batch_response_handlers[returned_name]
```

**Comment by pmarques** [comment]
This is working as expected now, thanks!

---
`ignore my previous comment (is now deleted), when tested the patch the first time I had a conflict that I missed. Sorted now!`

**Pull request merged**